src/main/java/spireMapOverhaul/zones/packattack/PackAttackZone.java

```diff
@@ -0,0 +1,179 @@
+                this.cards.addAll((ArrayList<AbstractCard>)ReflectionHacks.getCachedField(abstractCardPack, "cards").get(pack));
+            } catch (IllegalAccessException e) {
+                e.printStackTrace();
+                throw new RuntimeException("Error determining packs for Pack zone", e);
+            }
+        }
+        this.commonPool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.COMMON).collect(Collectors.toCollection(ArrayList::new));
+        this.uncommonPool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.UNCOMMON).collect(Collectors.toCollection(ArrayList::new));
+        this.rarePool = this.cards.stream().filter(c -> c.rarity == AbstractCard.CardRarity.RARE).collect(Collectors.toCollection(ArrayList::new));
+        this.updateDescription();
+    }
+
+    @Override
+    public float zoneSpecificEventRate() {
+        return 1;
+    }
+
+    @Override
+    public void atPreBattle() {
+        AbstractDungeon.player.gameHandSize -= 1;
+    }
+
+    @Override
+    public void atTurnStart() {
+        AbstractCard c = this.cards.get(AbstractDungeon.cardRandomRng.random(this.cards.size() - 1)).makeCopy();
+        Wiz.atb(new MakeTempCardInHandAction(c, 1));
+    }
+
+    @Override
+    public void modifyRewardCards(ArrayList<AbstractCard> cards) {
+        this.replaceWithPackCards(cards, AbstractDungeon.cardRng);
+        this.applyStandardUpgradeLogic(cards);
+    }
+
+    @Override
+    public void postCreateShopCards(ArrayList<AbstractCard> coloredCards, ArrayList<AbstractCard> colorlessCards) {
+        this.replaceWithPackCards(coloredCards, AbstractDungeon.merchantRng);
+        for (AbstractCard c : coloredCards) {
+            for (AbstractRelic relic : AbstractDungeon.player.relics) {
+                relic.onPreviewObtainCard(c);
+            }
+        }
+    }
+
+    private void replaceWithPackCards(ArrayList<AbstractCard> cards, Random rng) {
+        List<AbstractCard.CardRarity> validRarities = Arrays.asList(AbstractCard.CardRarity.COMMON, AbstractCard.CardRarity.UNCOMMON, AbstractCard.CardRarity.RARE);
+        for (int i = 0; i < cards.size(); i++) {
+            AbstractCard.CardRarity rarity = cards.get(i).rarity;
+            if (!validRarities.contains(rarity)) {
+                rarity = AbstractCard.CardRarity.COMMON;
+            }
+            cards.set(i, this.getPackCard(rarity, rng));
+        }
+    }
+
+    private AbstractCard getPackCard(AbstractCard.CardRarity rarity, Random rng) {
+        ArrayList<AbstractCard> options;
+        switch(rarity) {
+            case COMMON:
+                options = this.commonPool;
+                break;
+            case UNCOMMON:
+                options = this.uncommonPool;
+                break;
+            case RARE:
+                options = this.rarePool;
+                break;
+            default:
+                throw new RuntimeException("Invalid card rarity for rewards: " + rarity);
```

Comment:
I think by default we should just return a common card or something. Don't see a reason to make this an incompatibility

Reply:
That check is already in place, just that because of how I structured this it's in the calling method instead (`replaceWithPackCards`). Though looking at the code now, it's less code and more sensible for it to be in this method instead, so I'll do that. One of those things you don't always notice when writing the code.